sirepo/uri.py

```diff
@@ -58,14 +59,22 @@ def local_route(sim_type, route_name=None, params=None, query=None):
     s = simulation_db.get_schema(sim_type)
     if not route_name:
         route_name = default_local_route_name(s)
-    parts = s.localRoutes[route_name].route.split("/:")
+    n = "localRoutes"
```

Comment:
It seems localRoutes and reactRoutes are the same except for the fact that `source` is the only one different. I don't see a reason why source couldn't be the same for React. The SPA will have to redirect old URIs anyway so it could do this manually by changing the anchor form to the non-anchor form of the uri. Perhaps `/source/:sid` doesn't work right in React?

I think we only want to check `is_react_sim_type` and ignore whether the route is in reactRoutes. If it is not in reactRoutes, it still needs to be a localRoute. If React doesn't implement that route, it should either error or do something intelligent in the SPA. `uri` should only need to know that the route needs an anchor or not.

The code would look cleaner with a function like this in any event:

```py
    def _uri(route_table):
        parts = s[route_table][route_name].route.split("/:")
        u = parts.pop(0)
	for p in parts:
            if p.endswith("?"):
		p = p[:-1]
		if not params or p not in params:
                    continue
            u += "/" + _to_uri(params[p])
	return u
```

`route_table` might go away if my conjecture above is true.

Reply:
I was able to remove reactRoutes from schema-common.json and reverted the change in sirepo.uri. It still needs to use the # urls for angular and normal urls for react.

Reply:
Great, thanks.